Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,521 @@
+                        (Guid)ConnectionAfter.OperationId.Fetch(evnt.Value),
+                        (string)ConnectionAfter.Operation.Fetch(evnt.Value),
+                        (Guid)ConnectionAfter.ConnectionId.Fetch(evnt.Value),
+                        (SqlConnection)ConnectionAfter.Connection.Fetch(evnt.Value),
+                        (IDictionary)ConnectionAfter.Statistics.Fetch(evnt.Value),
+                        (long)ConnectionAfter.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlErrorOpenConnection:
+                case SqlErrorCloseConnection:
+                {
+                    this.OnErrorConnectionEvent(
+                        (Guid)ConnectionError.OperationId.Fetch(evnt.Value),
+                        (string)ConnectionError.Operation.Fetch(evnt.Value),
+                        (Guid)ConnectionError.ConnectionId.Fetch(evnt.Value),
+                        (Exception)ConnectionError.Exception.Fetch(evnt.Value),
+                        (long)ConnectionError.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlBeforeCommitTransaction:
+                case SqlBeforeRollbackTransaction:
+                {
+                    this.operationStartTimestamps.TryAdd(
+                        (Guid)TransactionBefore.OperationId.Fetch(evnt.Value),
+                        (long)TransactionBefore.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlAfterCommitTransaction:
+                case SqlAfterRollbackTransaction:
+                {
+                    this.OnAfterTransactionEvent(
+                        (Guid)TransactionAfter.OperationId.Fetch(evnt.Value),
+                        (string)TransactionAfter.Operation.Fetch(evnt.Value),
+                        (IsolationLevel)TransactionAfter.IsolationLevel.Fetch(evnt.Value),
+                        (SqlConnection)TransactionAfter.Connection.Fetch(evnt.Value),
+                        (long)TransactionAfter.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+
+                case SqlErrorCommitTransaction:
+                case SqlErrorRollbackTransaction:
+                {
+                    this.OnErrorTransactionEvent(
+                        (Guid)TransactionError.OperationId.Fetch(evnt.Value),
+                        (string)TransactionError.Operation.Fetch(evnt.Value),
+                        (IsolationLevel)TransactionError.IsolationLevel.Fetch(evnt.Value),
+                        (SqlConnection)TransactionError.Connection.Fetch(evnt.Value),
+                        (Exception)TransactionError.Exception.Fetch(evnt.Value),
+                        (long)TransactionError.Timestamp.Fetch(evnt.Value));
+
+                    break;
+                }
+            }
+        }
+
+        private static void InitializeTelemetry(ITelemetry telemetry, Guid operationId)
```

Comment:
@lmolkova Please check this. I may well have misunderstood your comments on error event correlation. Thanks!